### PR TITLE
[SPARK-47669][SQL][CONNECT][PYTHON] Add `Column.try_cast`

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Add `try_cast` function in Column APIs

### Why are the changes needed?
for functionality parity

### Does this PR introduce _any_ user-facing change?
yes

```
        >>> from pyspark.sql import functions as sf
        >>> df = spark.createDataFrame(
        ...      [(2, "123"), (5, "Bob"), (3, None)], ["age", "name"])
        >>> df.select(df.name.try_cast("double")).show()
        +-----+
        | name|
        +-----+
        |123.0|
        | NULL|
        | NULL|
        +-----+

```

### How was this patch tested?
new tests

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -1957,6 +1958,29 @@ object functions {
    */
   def try_sum(e: Column): Column = Column.fn("try_sum", e)
 
+  /**
+   * Casts the column to a different data type and the result is null on failure.
+   *
+   * @group normal_funcs
+   * @since 4.0.0
+   */
+  def try_cast(e: Column, dataType: DataType): Column = Column { builder =>
```
Comment (ueshin) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:1967]: I think we should have it as `Column` method same as `cast`?

File: python/pyspark/sql/functions/builtin.py [hunk 2]
```diff
@@ -20166,6 +20166,72 @@ def try_reflect(*cols: "ColumnOrName") -> Column:
...
 
 
+@_try_remote_functions
+def try_cast(
+    col: "ColumnOrName",
+    dataType: Union[DataType, str],
+) -> Column:
+    """
+    This is a special version of `cast` that performs the same operation, but returns a NULL
+    value instead of raising an error if the invoke method thrown exception.
```
Comment (xinrong-meng) [inline python/pyspark/sql/functions/builtin.py:20176]: nit: thrown -> throws

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (zhengruifeng) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:1967]: sure, let me change it to `Column.try_cast`

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 3]
```diff
@@ -1222,6 +1222,43 @@ class Column(val expr: Expression) extends Logging {
...
+   *
+   * @group expr_ops
+   * @since 4.0.0
+   */
+  def try_cast(to: DataType): Column = withExpr {
+    val cast = Cast(
+      child = expr,
+      dataType = CharVarcharUtils.replaceCharVarcharWithStringForCast(to),
+      evalMode = EvalMode.TRY)
+    cast.setTagValue(Cast.USER_SPECIFIED_CAST, ())
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala:1244]: Just curious, what is this tag doing?

**Review by ueshin — APPROVED** [review]
LGTM.
I just left a couple questions.

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 4]
```diff
@@ -137,6 +137,13 @@ message Expression {
     }
   }
 
+  enum EvalMode {
+    EVAL_MODE_UNSPECIFIED = 0;
+    EVAL_MODE_LEGACY = 1;
+    EVAL_MODE_ANSI = 2;
+    EVAL_MODE_TRY = 3;
+  }
```
[the remark this answers is not in the document]
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:145]: good point, let me move it there

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala (hunk 3 above)
Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/Column.scala:1244]: https://github.com/apache/spark/blob/f18d945af7b69fbc89b38b9ca3ca79263b0881ed/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L182-L185

I guess it is used to distinguish between user calls and internal calls.
Seems we never support it in Spark Connect though

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/Column.scala:1244]: this tag was introduced in https://github.com/apache/spark/commit/92e5cfd58dc7dd5a5b1d3e2b6e93a03af5a07b67, seems it affects the `sql` string and `prettyName`

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/Column.scala:1244]: it seems this tag was used to change the field name in sql text `spark.sql("select floor(1)")`

I suspect it doesn't take affect in dataframe API, I compared the output column name of `cast/try_cast`, and the classic and connect are the same.

**Review by xinrong-meng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
thank you all
merged to master

**Pull request closed without merging**